### ISSUE TITLE
Plugins and the representation of positions (utf8 and utf16, absolute and line/column)

### ISSUE THREAD

**Issue by cmyr** [google/xi-editor#746]
## Represent positions

We currently have two principal ways of representing positions in a buffer; either as absolute utf-8 offsets, or as a line number + a utf-8 offset column position.

The Language Server Protocol uses utf-16 line + column pairs to represent all positions. To talk to language servers, we _have_ to convert whatever representation we're using into utf-16 line/col. This is true for things like the position of click events, but it's also true for things like the offsets in a delta.

### From core to plugins
There are a number of ways we could handle this:
#### Convert in the plugin

We could continue sending most locations to plugins as utf-8 offsets, and leave it up to the plugin to handle the conversion. This keeps core conceptually simple, but complicates the writing of plugin libs; almost all plugin libs are going to want to be able to convert offsets to line/col positions of _some_ kind, and so this work will be duplicated by every plugin library.

#### Convert in core, per plugin

We could allow each plugin to specify how it would like to receive locations. Before sending a message to a plugin, we could convert any position information in that message to the plugin's desired format.

#### Convert in core, aggressively

We could _always_ send _all_ representations of a position to the plugin. That is, plugins could receive positions as 4 separate offsets; the absolute utf-8 offset, the line number, and the column in both utf-8 and utf-16 unit spaces. (I'm not sure there is an argument for including the absolute utf-16 offset or not.)

This will minorly increase the size of the RPC messages, but will also remove the need for any conversion to happen in the plugin libs, in the general case. It also has the advantage of only requiring these conversions to be done once per event, and they can be done performantly in core. It will also be easier to write unit tests for the conversion, and if any plugins need to implement their own conversion logic they will have a good source of validated test data.

It is also... inelegant.

### Going the other way (plugins to core)

What do we do with plugins that _send_ us positons in utf-16? This is likely only a concern when interfacing with external code, as with language servers.

If we've chosen to do all the original conversions in the plugin lib, then it makes sense that we would do _these_ conversions there as well. If we're doing conversion in core, though, this gets a bit ugly:

If we're converting positions in core when we send them _to_ plugins, do we allow plugins to send us positions in various coordinate spaces, and normalize them in core? This _also_ makes core a bit messier, but seems to be a direct consequence of converting in core on the way out; if we want to keep conversion logic in one place, we need to do that fully.

## So...

I think the idea of converting aggressively in core makes a lot of sense, but I _do_ dislike the idea of allowing plugins to send positions back to core in multiple formats. We need to accept one side or the other of this compromise. Which should it be? 😕

**Comment by betterclever** [comment]
I tried to do all the conversions in the Plugin while cleaning #717 for Hover only. And it feels like I ran into same problem all over again. We don't have view in scope at the time we receive the result from the Language Server, it is delivered asynchronously to other thread. This was one of the reason that we thought about doing conversions in the core at the first place.

An alternative that I could think of, is to pre-fetch whole document while making the request from the Hover in the LSP Plugin then we can do the conversion even after the result, but it look too artificial to me. At this point, I am totally confused what should be the correct step forward.

**Comment by cmyr** [comment]
Okay, so problems with doing conversion in plugin:

- plugin may not have the correct rev
- postion to be converted may be in file that is not open in editor
- architecture of lsp lib means views are not in scope when response is received

Of these, the first two seem the hardest; the third one is something we could address with enough work.

The second of these, in particular, seems intractable without core involvement.

If we're going to do this in core, I would like to do this _all_ in core; that is, both sending and receiving.

For sending, I'm inclined towards the "fat position" approach, that is sending the position in multiple representations.

For receiving, we will have a `Position` enum, that will be used in RPC; internally we will convert to utf-8 offsets as close to the RPC boundary as possible. The `Position` type should only ever be seen by code that initially handles the RPC.

So for hover, e.g., the type on the wire might be:

```js
hover_item { "text": String, "start": Position?, "end": Position? }
```

but in core, this would be converted to something like:

```rust
Struct HoverItem {
    text: String,
    start: Option<usize>,
    end: Option<usize>,
}
```

Basically: we want the implementation detail of accepting various position representations touch as little core code as possible.

**Comment by betterclever** [comment]
I am also in for the approach exactly as proposed here. Though, out of curiosity, if we were still doing this in Plugin, what could be an alternative way to get the document content to do conversions when the result return from the Language Server.

**Comment by cmyr** [comment]
I was thinking we could do something like keep track of which view originated each pending request, and when the response arrived we could grab the relevant view and pass it to a closure supplied by the caller.
